bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/discovery/VeluxBridgeFinder.java

```diff
@@ -0,0 +1,315 @@
+
+                    // parse the answers
+                    for (short an = 0; an < anCount; an++) {
+                        // parse the name
+                        byte[] str = new byte[BUFFER_SIZE];
+                        int i = 0;
+                        int segLength;
+                        while ((segLength = dataStream.readByte()) > 0) {
+                            for (int index = 0; index < segLength; index++) {
+                                str[i] = dataStream.readByte();
+                                i++;
+                            }
+                            str[i] = '.';
+                            i++;
+                        }
+                        String name = new String(str, 0, i, StandardCharsets.UTF_8);
+                        short typ = dataStream.readShort();
+                        short clazz = (short) (CLASS_MASK & dataStream.readShort());
+                        if (!(name.startsWith(KLF_SERVICE_ID)) || (typ != TYPE_PTR) || (clazz != CLASS_IN)) {
+                            return false;
+                        }
+
+                        // if we got here, the name and response type are valid
+                        @SuppressWarnings("unused")
+                        int ttl = dataStream.readInt();
+                        @SuppressWarnings("unused")
+                        short dataLen = dataStream.readShort();
+
+                        // parse the host name
+                        i = 0;
+                        while ((segLength = dataStream.readByte()) > 0) {
+                            for (int index = 0; index < segLength; index++) {
+                                str[i] = dataStream.readByte();
+                                i++;
+                            }
+                            str[i] = '.';
+                            i++;
+                        }
+
+                        // check if the host name matches
+                        String host = new String(str, 0, i, StandardCharsets.UTF_8);
+                        if (host.startsWith(KLF_HOST_PREFIX)) {
+                            return true;
+                        }
+                    }
+                }
+            }
+        } catch (IOException e) {
+            // fall through
+        }
+        return false;
+    }
+
+    /**
+     * Start the listener task
+     */
+    private void startListener() {
+        listenerTask = scheduler.submit(listenerRunnable);
+    }
+
+    /**
+     * Stop the listener task
+     */
+    private void stopListener() {
+        Future<?> listenerTask = this.listenerTask;
+        if (listenerTask != null) {
+            listenerTask.cancel(false);
+        }
+        this.listenerTask = null;
+    }
+
+    /**
+     * Private method that searches for Velux Bridges and returns their IP addresses
+     *
+     * @return list of dotted IP address e.g. '123.123.123.123'
+     */
+    private Set<String> discoverBridgeIpAddresses() {
```

Comment:
Since this is a very long running task you should run this asynchronously so that it can be cancelled by the caller if they need to.

```suggestion
    private Future<Set<String>> discoverBridgeIpAddresses() {
```

As far as how to implement the future, I suggest using a FutureTask.

Reply:
Thanks for the suggestion. I think the simplest solution is to wrap the whole thing in a lambda expression called on the discovery service's task scheduler.